tls/s2n_early_data.c

```diff
@@ -109,9 +116,17 @@ S2N_RESULT s2n_early_data_accept_or_reject(struct s2n_connection *conn)
 
     /* Even if the connection is valid for early data, the client can't consider
      * early data accepted until the server sends the early data indication. */
-    if (conn->mode == S2N_SERVER) {
-        RESULT_GUARD(s2n_connection_set_early_data_state(conn, S2N_EARLY_DATA_ACCEPTED));
+    if (conn->mode == S2N_CLIENT) {
```

Comment:
What if the client's early_data_expected bitflag is false? Shouldn't this method not just return but also change the early data state to S2N_EARLY_DATA_REJECTED?

Reply:
No. The client can't arbitrarily decide that early data is rejected. It can only mark early data as rejected if the connection is in such a state that even if the server _did_ eventually respond with "accepted", that would be a protocol error. If the client requests early data, then marks it "rejected" because of application input, how would it communicate that decision to the server? Because if the server accepts early data, it's going to expect an EndOfEarlyData message, which isn't sent when early data is rejected.

The client's early_data_expected bitflag only comes into play when requesting early data. After that, it's out of the client's hands: we can send zero application data if we want to, but we MUST send the EndOfEarlyData message.